src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/StockInMemoryRepositoryImpl.kt

```diff
@@ -1,52 +1,100 @@
-                0.5
-            ))
-        }.take(5)
+        val maxPercentChange = getMaxBigDecimal(Stock::percentChange)
+        val maxChange = getMaxBigDecimal(Stock::change)
+
+        val timeOfRequest = Instant.now()
+
+        return db.asSequence()
+            .filterNot { (_, stock) -> stock.symbol == null }
+            .filter { (_, stock) ->
+                isStockInValidDateRange(stock, timeOfRequest)
+            }
+            .map { it.value }
+            .groupBy { it.symbol!! }
+            .toList()
+            .sortedByDescending(selectorFunctionToCompareStocksByPrice(maxChange, maxPercentChange))
+            .take(n)
             .map { (symbol, stocks) ->
                 symbol to stocks.distinctBy { it.currentPrice }
-                    .sortedByDescending { stock -> stock.dateOfRetrieval }.take(5)
+                    .sortedByDescending { it.dateOfRetrieval }
+                    .take(NUMBER_OF_HISTORY_RECORDS_PER_STOCK)
             }
+            .toList()
+    }
+
+    private fun selectorFunctionToCompareStocksByPrice(
+        maxChange: BigDecimal,
+        maxPercentChange: BigDecimal,
+    ): (Pair<String, List<Stock>>) -> BigDecimal {
+        return { pair: Pair<String, List<Stock>> ->
+            val stockList = pair.second
+            val avgChange = getAvgOfBigDecimals(stockList) { stock: Stock -> stock.change }
+            val avgPercentChange = getAvgOfBigDecimals(stockList) { stock: Stock -> stock.percentChange }
+
+            val priceChangeCoefficient =
+                (avgChange / maxChange) * BigDecimal.valueOf(WEIGHTED_COEFFICIENT_FOR_PRICE_COEFFICIENTS)
+            val pricePercentChangeCoefficient =
+                (avgPercentChange / maxPercentChange) * BigDecimal.valueOf(WEIGHTED_COEFFICIENT_FOR_PRICE_COEFFICIENTS)
+            priceChangeCoefficient + pricePercentChangeCoefficient
+        }
+    }
+
+    private fun isStockInValidDateRange(stock: Stock, timeOfRequest: Instant): Boolean {
+        return stock.dateOfRetrieval
+            ?.isBefore(
+                timeOfRequest.plus(Duration.ofHours(1))
+            ) == true
+    }
+
+    private fun getMaxBigDecimal(property: KProperty1<Stock, BigDecimal?>): BigDecimal {
+        return db
+            .mapNotNull { (_, stock) -> property.get(stock) }
+            .maxBy { it }
+    }
+
+    private fun getAvgOfBigDecimals(
+        stockList: List<Stock>,
+        mapperToBigDecimal: (stock: Stock) -> BigDecimal?,
+    ): BigDecimal {
+        return stockList
+            .mapNotNull { mapperToBigDecimal(it) }
+            .fold(BigDecimal.ZERO) { acc, bigDecimal -> acc + bigDecimal } /
+                BigDecimal.valueOf(stockList.size.toLong())
     }
 
-    override fun getAllStockSymbols(): List<String?> {
-        return db.asSequence().map { it.value.symbol }.distinct().toList()
+    override fun findAllStockSymbols(): List<String> {
```

Comment:
```suggestion
    override fun findAllStockSymbols(): Set<String> {
```
don't see reason to convert list to set and again to list